### PR TITLE
Add Typescript typechecking, generate types with tsc

### LINKED ISSUE #152: Documentation of the [lat, lng] tuple
Many methods (e.g. `cellToLatLng()`) return a tuple of latitude, longitude.

They are documented as returning a variable-length array of numbers (`number[]`), while in practice they should return a fixed-length array.

Is there a reason not to document this as `[number, number]` instead? This could help in TS projects with [strictly configured index access](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess).

Thanks

Further reading
* there seems to have been a [discussion around this in the past](https://github.com/uber/h3-js/pull/91#discussion_r487145676)
* https://stackoverflow.com/questions/64739315/how-to-document-tuple-using-jsdoc

### PR DESCRIPTION
**Pull request by nrabinowitz**
- Adds type-checking via `tsc --noEmit`
- Updates types in JSDoc comments to fix TS errors
- Drops `tsd-jsdoc` in favor of generating the library types with `tsc --declaration`
- Adds types for `CoordPair` and `SplitLong`, post-processing the libdef to use `[number, number]` instead of `number[]` (this is a workaround for https://github.com/jsdoc/jsdoc/issues/1703, which kills doc generation if we use `[number, number]` directly in JSDoc comments)

Fixes #152

### CODE REVIEW

**Comment by isaacbrodsky** [comment]
Can this be merged after v4.0.0 is released in both core library & JS?

**Comment by nrabinowitz** [comment]
> Can this be merged after v4.0.0 is released in both core library & JS?

Technically changing types could be considered breaking for TS users, so I'd prefer to land this pre-release.

File: package.json [hunk 1]
```diff
@@ -30,7 +30,8 @@
     "build-emscripten": "yarn build-update-h3 && yarn docker-emscripten",
     "build-legacy": "node scripts/build-legacy-api.js",
     "build-docs": "jsdoc2md --no-cache --global-index-format grouped --partial doc-files/scope.hbs --helper ./doc-files/insert-version.js --separators --template doc-files/README.tmpl.md lib/h3core.js lib/errors.js > README.md",
-    "build-tsd-core": "jsdoc -t node_modules/tsd-jsdoc/dist -d console lib/h3core.js | sed 's/\"h3\"/\"h3-js\"/g' > dist/types.d.ts",
+    "build-tsd-core": "tsc lib/h3core.js --noResolve --skipLibCheck --allowJs --declaration --emitDeclarationOnly --outFile dist/types.d.ts && yarn build-tsd-core-postprocess",
+    "build-tsd-core-postprocess": "sed 's/module \"h3core\"/module \"h3-js\"/g' dist/types.d.ts > types.d.ts.tmp && sed -E -e 's/(CoordPair|SplitLong) = number\\[\\]/\\1 = [number,number]/g' types.d.ts.tmp > dist/types.d.ts && rm types.d.ts.tmp",
```
Comment (isaacbrodsky) [inline package.json:34]: Maybe this logic should be in a script file so that comments can be added to it? This is a little 'magical'

**Review by isaacbrodsky — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (nrabinowitz) [reply package.json:34]: Agree, I can take a look at moving this into a script

**Pull request merged**